opentelemetry/src/global/trace.rs

```diff
@@ -285,9 +268,276 @@ where
+            }
+
+            fn flush(&mut self) -> std::io::Result<()> {
+                let mut buffer = self.buf.lock().expect("cannot acquire the lock of assert writer");
+                buffer.flush()
+            }
+        }
+
+        impl Clone for AssertWriter {
+            fn clone(&self) -> Self {
+                AssertWriter {
+                    buf: self.buf.clone()
+                }
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    struct TestTracerProvider {
+        debug_msg: &'static str,
+    }
+
+    impl Default for TestTracerProvider {
+        fn default() -> Self {
+            TestTracerProvider { debug_msg: "" }
+        }
+    }
+
+    impl TestTracerProvider {
+        fn new(debug_msg: &'static str) -> Self {
+            TestTracerProvider { debug_msg }
+        }
+    }
+
+    impl TracerProvider for TestTracerProvider {
+        type Tracer = NoopTracer;
+
+        fn get_tracer(&self, _name: &'static str, _version: Option<&'static str>) -> Self::Tracer {
+            NoopTracer::default()
+        }
+    }
+
+    #[test]
+    #[ignore]
+    fn test_set_tracer_provider() {
+        let _guard = set_tracer_provider(TestTracerProvider::new("global one"));
+
+        {
+            let _guard = set_tracer_provider(TestTracerProvider::new("inner one"));
+            assert!(format!("{:?}", tracer_provider()).contains("inner one"));
+        }
+
+        assert!(format!("{:?}", tracer_provider()).contains("inner one"));
+    }
+
+    #[test]
+    #[ignore]
+    fn test_set_tracer_provider_in_another_thread() {
+        let _guard = set_tracer_provider(TestTracerProvider::new("global one"));
+
+        let handle = thread::spawn(move || {
+            assert!(format!("{:?}", tracer_provider()).contains("global one"));
+        });
+
+        println!("{:?}", tracer_provider());
+
+        let _ = handle.join();
+    }
+
+    #[test]
+    #[ignore]
+    fn test_set_tracer_provider_in_another_function() {
+        let setup = || {
+            let _guard = set_tracer_provider(TestTracerProvider::new("global one"));
+            assert!(format!("{:?}", tracer_provider()).contains("global one"))
+        };
+
+        setup();
+
+        assert!(format!("{:?}", tracer_provider()).contains("global one"))
+    }
+
+    #[test]
+    #[ignore]
+    fn test_set_two_provider_in_two_thread() {
+        let (sender, recv) = std::sync::mpsc::channel();
+        let (sender1, sender2) = (sender.clone(), sender);
+        let _handle1 = thread::spawn(move || {
+            sleep(Duration::from_secs(1));
+            let _guard = set_tracer_provider(TestTracerProvider::new("thread 1"));
+            sleep(Duration::from_secs(2));
+            let _ = sender1.send(format!("thread 1: {:?}", tracer_provider()));
+        });
+        let _handle2 = thread::spawn(move || {
+            sleep(Duration::from_secs(2));
+            let _guard = set_tracer_provider(TestTracerProvider::new("thread 2"));
```

Comment:
This no longer returns a guard. Should we rename the variable to sth like `_previous` or just `_`?